### ISSUE TITLE
Compress marks in memory.

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#45941]
Marks are represented as an array of two 64-bit numbers: `offset_in_compressed_file`, `offset_in_decompressed_block`.
The array requires only one operation after it is loaded - random read access.

- `offset_in_compressed_file` is monotonically non-decreasing, usually by a similar value;
- `offset_in_decompressed_block` is zero most of the time, but sometimes not.

Let's implement ad-hoc compression for this array.

It will be split into 1024-element blocks, and every block will contain the following:

- minimum value of `offset_in_compressed_file`;
- the number of bits required to store the difference between the maximum and minimum value of `offset_in_compressed_file`;
- minimum value of `offset_in_decompressed_block` when it is non-zero;
- the number of bits required to store the difference between the maximum and minimum non-zero value of `offset_in_decompressed_file` (can be 0 if all of the values are 0);

Then we will represent every `offset_in_compressed_file` as the difference with the minimum value, bit packed,
and every `offset_in_decompressed_block` as one bit indicating if it is zero, and other bits as the difference with the minimum non-zero value, bit packed.

Marks will always be compressed in memory, unconditionally.
The expected compression ratio will be around 6 times.

See also #30434

**Comment by al13n321** [comment]
Sounds good!

1024 sounds reasonable, but I wondered how much it can be improved and ended up with the calculation below.

Tl;dr: optimum seems to be 177-element blocks, and which is ~5% more compact than 1024-element blocks (*).

Let D - typical difference between consecutive offset_in_compressed_file values, B - marks block size (1024 in the OP), 32 bytes - fixed overhead per block. Then this data structure will use about this many bits per mark:
```
32*8/B + log_2(D*B) -> min
```

Optimal B doesn't depend on D (!) and is just `32 * 8 * log(2)`, apparently.

(*) If D (the typical difference between consecutive offset_in_compressed_file values) is 10k then 177 is ~5% better than 1024. The "5%" increases if D decreases.

**Comment by alexey-milovidov** [comment]
Ok. Maybe set it to 256, so people wouldn't bother thinking if integer division to a constant is fast or not...

**Comment by al13n321** [comment]
Can also make the differences be relative to a linear interpolation rather than a single per-block min value. I.e. one of:
 * Make the block have first value and step size for offset_in_compressed_file. The compressed marks would be signed integers. The i'th mark is `base + step * i + compressed_value`.
 * Or instead of explicitly storing the step we can interpolate between current and next block's base offset.

Not sure yet whether this is worth the trouble. Maybe I'll try both and see how much code this adds.

**Comment by al13n321** [comment]
Hm, in compact format, marks of all columns are interleaved. For compression, I guess we have to either un-interleave them (and have worse cache locality when accessing multiple columns?) or make each block have a whole array of min offsets and bit widths. Or we could yolo it and use the same min offset for all columns, but that seems too wasteful.

I'll make MergeTreeMarksLoader un-interleave the marks and keep a separate MarksInCompressedFile for each column.

**Comment by alexey-milovidov** [comment]
We expect that there is only a small amount of data in compact parts, so the marks of compact parts can remain uncompressed.

**Comment by alexey-milovidov** [comment]
Uninterleaving is also ok.

PS. I thought the format in memory is the same for different part formats, looks like it's not the case.

**Comment by al13n321** [comment]
(Nevermind everything about un-interleaving, I misunderstood how that works.)
